README

This repostiroy does NOT contain an official version of squidGuard (see below)

The official squidGuard homepage is:

		      http://www.squidguard.org/

			      What it is
			      ~~~~~~~~~~

squidGuard is a free (GPL), flexible and ultra fast filter, redirector
and access controller plugin for squid.  It lets you define multiple
access rules with different restrictions for different user groups on
a squid cache.  squidGuard uses squid's standard redirector interface.


			       Authors
			       ~~~~~~~

The initial squidGuard concept was designed by Pål Baltzersen and was
implemented and was maintained and extended by Lars Erik Håland at
ElTele Øst AS. 
Since December 2006 squidGuard is maintained by Shalla Secure Services.


			     Distribution
			     ~~~~~~~~~~~~

squidGuard is distributed by Shalla Secure Services under GPLv2 and may
therefore be freely used and distributed according to the conditions of
the licence.


========================================================================

The version here is a rewrite of most of the code. While it provides (almost)
all features found in SG 1.5, the configuration is not entirely compatible.

Please read on ...


New Features (the good news)
============================

Authorization Helper for Squid
------------------------------

SG can now be used as an external authorization helper for squid. This mode
requires a particular Squid setup, for example

  external_acl_type sgaccess  ttl=300 children=8\
           %URI %SRC %un /usr/bin/squidGuard -d -z -c /etc/squidGuard.conf
  acl sgaccess external sgaccess
  http_access allow sgaccess
  http_access deny all

The new "tag" statement within an ACL statement can be used to pass a tag back
to squid, which can then subsequently be used in further Squid-ACLs or in log
statements.

Of course, "redirect" and "rewrite" statements in the configuration won't work
in this mode.

Note: to use "%un" as above with Squid 3.1 requires a patch for Squid, see

  http://bugs.squid-cache.org/show_bug.cgi?id=3345

Alternatively, you can use "%LOGIN" instead, but then you can not have optional
authentication.

IPv6 Support
------------

In source blocks, IPv6 addresses can be used just like IPv4 addresses, except
that you need to quote them. To match all private address ranges:

  source private {
     ip "fc00::/7" "fe80::/10"
     ip "10.0.0.0/8" "172.16.0.0/12" "192.168.0.0/16"
  }

The built-in destination list "in-addr" now matches IPv6 addresses in URLs, such
as in "http://[2a02:2e0:3fe:100::7]".

IPv6 addresses in URL/domain lists won't work very well though: There is some
variation how one could write the same address and SG currently does not
translate the IP into a canonical form.

Reverse DNS lookups for URLs
----------------------------

Reverse lookups can be enabled for IP addresses in URLs. This works for IPv6
and IPv4 addresses.

System user group support
-------------------------

A new "group" statement is available in "source" blocks to check if a user is member
of a system group. In conjunction with Samba, this also enables AD group support.

"netgroup" support for users
----------------------------

Users can now also be checked against membership in a NIS netgroup.

There is currently no support to check against a specific NIS domain and no support
for IP address checks, but that would now be easy to implement.


Configuration Syntax Changes
============================

General
-------

While unquoted strings still work almost anywhere, you should always quote
words that are not supposed to be interpreted as a keyword. For example, use

  user "fred" "barney" "mon"

instead of

  user fred barney mon

The latter would result in a syntax error, because the unquoted "mon" is
interpreted as the weekday "monday". This is no new problem though and you
can use quoted strings since 1.5 or so.

Settings
--------

Global settings now need to be prefixed with the "set" keyword,
for example

  set dbhome "/var/lib/squidGuard"
  set logdir "/var/log/squidGuard"

instead of

  dbhome /var/lib/squidGuard
  logdir /var/log/squidGuard

Due to this change, global settings are no longer keywords and it is now much
easier to introduce new such settings without causing surprises with existing
configurations.

Default Redirect
----------------

A "redirect" statement outside of any block is used as a default redirect when
no more specific redirect or rewrite rule can be found. This is especially
useful in cases where you want to redirect all blocked requests to a specific
page or CGI and do not want to repeat the same redirect statement over and over
again.


Request Logs
------------

A separate log can be specified per ACL, per source, per destination or
globally (with this precedence).

Logs in a rewrite block are still accepted by the configuration parser but are
silently ignored.

The global request log file must now be specified with a "logfile" directive
outside any block.

There is still no configurable log format and you can still not log to anything
but files.

"time" blocks
-------------

Not much intentional changes here, except that a time specification now
includes the final minute of the range, hence "00:00-23:59" really means the
whole day.

"source" blocks
---------------

* quotas have been removed.

* the behavior of the "continue" statement is now implied and the word is
  no longer a keyword (= using it is a syntax error).

IP addresses:

  * the "ip" keyword now allows IPv6 addresses.

  * IP masks can no longer be specified in dot notation, f.e
    192.168.1.0/255.255.255.0 must now be written as 192.168.1.0/24

  * IP ranges are no longer supported. Use CIDR notation instead.

  * the same rules apply to address list read from a file.

External lookups

  * live-lookups are executed in the order specified.

  * all lookups are cached. You can specify the cache time with the new
    setting "source-ttl". The special settings for "ldapusersearch" and
    "ldapipsearch" have been removed.

  * "ip", "user", "iplist" and "userlist" statements just populate the
    lookup cache permanently, they are no live lookups.

  * MySQL support has been removed. The old code did not actually
    compile and the thing it did was a bit stupid. On the bright side, a
    properly working live-lookup can now be implemented much more easily.

"destination" blocks
--------------------

* DNS blacklists can be specified with the "dnsbl" keyword within a destination
  block. For example

  destination "uribl" {
    dnsbl ".black.uribl.com"
  }

  If you use DNS blacklists, you want to run a local DNS server (preferably on
  the local host) for caching.

ACLs
----

* DNS blacklists must now be specified in a destination block. See above.

* SG no longer searches the best matching source block for a request. Instead,
  ACLs are evaluated in the order specified.

* multiple ACLs for the same source are now possible.

* the "default"-ACL is no longer special and especially the behavior to take
  the default request log location from there no longer works. Otherwise, since
  an existing configuration file would not have a source "default" and a
  non-existing source matches any source, a "default" ACL as last rule would
  still work as before.

* the "else" keyword just creates an ACL with the same source as the previous
  one. You can also use multiple "else" blocks, with different time ranges for
  example.

* the new "next" keyword is similar to "pass", except that "pass" terminates
  ACL processing with a positive result, while "next" causes ACL processing to
  continue.

* the "allow" keyword is the opposite of redirect, meaning access is granted
  when the ACL matches.

  You can use "allow" in authorization helper mode if you just want to tag
  matching connection. The tag can then be used in the Squid configuration to
  change its behavior based on SG rules.

  In both modes "allow" is useful in conjunction with "next", for example

  acl {
     default {
        next restricted
	allow
     }
     trusted {
        pass all
     }
     suspect {
        pass none
     }
   }

* "tag" allows you to pass a tag back to squid, which can then be used in Squid
  access rules and log files.
  Tags only work in authorization helper mode since squid currently does not
  support it with a redirector.

Reasoning
=========

Recently, I needed some features that popped up in SG 1.4 and we had some
extensions based on the 1.2 version I had to port. Getting these extensions to
work again, was not easy and no fun at all.

And then I needed the LDAP support from 1.5-beta and had to port our stuff
again.

And then I wanted IPv6. And I also needed the authorization helper support to
pass informations back to Squid ...

While digging through the code, I finally realized, that SG truly was the
manifestation of the Flying Spaghetti Monster in C !

So, instead of converting to Pastafarism, I started cleaning up:

General code cleanup
--------------------

Among the changes was getting rid of K&R support. Heck, C90 was hot back then,
when I started serious software development. Nowadays, if your compiler does
not grasp this code, upgrade your compiler.

The code has been restructured in a way so that the C compiler can better
optimize (use of "static" functions) and can better detect possibly dangerous
constructs (use "const" when sensible, better use of function prototypes, use
-Wall -Werror). It has also been split in more isolated modules where only the
absolute necessary guts are exposed to other modules.


Better extensibility
--------------------

It is now significantly easier to add new source- or destination matches, and

CHANGELOG
Release 1.5
2010-09-09	Fixed inconsistent blocking (bug 59).  Replaced defined routine
		in sgDB.c
2010-09-08	Added Russian translation from Vladimir Ipatov to squidGuard.cgi.in.
2009-10-19	Fixed two bypass problems with URLs which length is close to 
		the limit defined by MAX_BUF. The resulting proxy line exceeds
		this limit and causes either squid or squidGuard to properly
		block a site.
2009-10-15	Fixed a problem with very long URLs. SquidGuard will go into 
		emergency mode when a overlong URLs are encountered. The
		emergency mode causes an entire stop of blocking. This is not
		appropriate in this situation.
2009-09-30	Added patch by beber and gentoo (thank you!) to fix a 
		problem when cross compiling (bug 56).
2009-09-27	Added patch by gentoo to fix alocal warnings (bug 57).
2009-09-15	Added a feature to send log messages to syslog based on the
		patch from Jun Jiang (thank you). (bug 42)  In order to use
		syslog you have to run configure with the new option
		"--with-syslog". In the configuration file you need to add
		a line "syslog enable". If any other value but "enable" is
		used syslog is disabled and logging to squidGuard.log takes
		place as usual.
		The following log level are used: DEBUG, NOTICE, WARN, ERROR
		and EMERG. The local4 syslog facility is used by default.
		If you want to change this, use the configure option
		"--with-syslog-facility=<facility>".
2009-09-12	Anonymized passwords (for connecting to the ldap or mysql 
		server) written to logfiles when squidGuard is starting.
		Added two configure options for choosing different location
		for the LDAP include and library files.
2009-08-25	Added patch to check IP addresses against LDAP. Patch by
		Denis Bonnenfant (bug 41) - thank you.
2009-08-23	Added patch to allow quoted strings in the configuration
		file (bug 53). For more information see README.QuotedStrings.
		Thanks to Iain Fothergill for providing the patch.
		Removed the fix for usernames starting with a number because
		it breaks the time declarations.
2009-05-08	Added patch by INL to enable blocking against DNS based
		blacklists (bug 55).
		Fixed re-opened bug 12: a problem with regular expressions.
		An entry like "www\.google\.de" did not block www.google.de
		which it was supposed to do. Solving this issue solved
		bug 46 as well.
2009-03-08	Fixed bug 52: Sometimes squidGuard crashes with an overflow
		error message for vsprintf. Thanks to Dirk Schoebel for 
		suggesting the proper fix.
		Fixed bug 49: Using numeric username made squidGuard goes
		into emergency mode. This has been fixed. Usernames can
		now start with a number, be numeric and can additionally
		contain the following characters: @,à,é,è,ñ,á,ì,í,ò,ó,ù,ú.


Release 1.4
2008-12-23	Included last fixes for 1.4 final.
2008-07-23	Some cleanup and fine work: added information about "-b"
		parameter to the help output. Added "!" to the list of
		allowed characters in urls.
2008-07-14	Fixed bug40: When an url ended in "://" squidGuard crashed.
		This has been fixed (the end of the url is now tested).
		Fixed bug39: squid was complaining about stdout messages from 
		squidGuard during the db update as a result of the progress
		bar output. The output has been moved to stderr.
2008-07-11	Added a switch to turn on the progress bar. The old default
		behaviour has been restored to not show anything. To see the
		progress of the compilation of the db files, use the command
			squidGuard -b
		(-b for "bar" - the progressbar).
2008-07-05	Fixed bug 37: The sample block script squidGuard.cgi has been
		rewritten to fix some bugs (basically warnings due to old perl
		syntax but problem with the language selection as well) and to
		enable an easier integration of additional languages. If a new
		language shall be supported create a babel file the same way
		the others are build and add the language to the supported hash
		in squidGuard.cgi. The location of the babel files is the same
		as for squidGuard.cgi. If you want to change the location change
		the path in squidGuard.cgi (line 298).
		The layout of the block page has been improved a bit as well.
		Fixed bug 23: Syntax error when using "~" in redirect statement.
		The "~" was not in the list of allowed characters for the
		redirect url.
		Fixed bug 34: On some system y.tab.c and y.tab.h were not properly
		created although bison was installed. This led to errors during
		compilation. If now no _executable_ bison is found, make copies
		the files over. The prepared files have been updated, too.
2008-06-13	Fixed bug 38: Bypass vulnerability using trailing dots. This 
		problem only occurs when squid 3.0 is used. Squid 2.6 removes
		trailing dots from domains before passing them to squidGuard.
2008-05-17	Removed the automatic download of the documentation when running

src/sgDiv.c
/*
 * By accepting this notice, you agree to be bound by the following
 * agreements:
 *
 * This software product, squidGuard, is copyrighted (C) 1998-2009
 * by Christine Kronberg, Shalla Secure Services. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License (version 2) as
 * published by the Free Software Foundation.  It is distributed in the
 * hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License (GPL) for more details.
 *
 * You should have received a copy of the GNU General Public License
 * (GPL) along with this program.
 */

#include "sg.h"
#include "sgEx.h"
#include "sgLog.h"
#include "sgMemory.h"
#include "HTEscape.h"

#include <netdb.h>
#include <arpa/inet.h>

/* #define METEST 8; */
int reverselookup = 0;

void sgHandlerSigHUP(int signal)
{
	sig_hup = 1;
}

void sgReloadConfig()
{
	struct LogFileStat *sg;
	struct Source *src;
	struct Destination *dest;

	sig_hup = 0;
	sgLogWarn("WARN: Received sigHUP, reloaded configuration");
	for (sg = LogFileStat; sg != NULL; sg = sg->next) { /* closing logfiles */
		if (sg->fd == stderr || sg->fd == stdout)
			continue;
		fclose(sg->fd);
	}
	for (src = Source; src != NULL; src = src->next) {
		if (src->domainDb != NULL && src->domainDb->dbp != NULL)
			(void)src->domainDb->dbp->close(src->domainDb->dbp, 0);
		if (src->userDb != NULL && src->userDb->dbp != NULL)
			(void)src->userDb->dbp->close(src->userDb->dbp, 0);
	}
	for (dest = Dest; dest != NULL; dest = dest->next) {
		if (dest->domainlistDb != NULL && dest->domainlistDb->dbp != NULL)
			(void)dest->domainlistDb->dbp->close(dest->domainlistDb->dbp, 0);
		if (dest->urllistDb != NULL && dest->urllistDb->dbp != NULL)
			(void)dest->urllistDb->dbp->close(dest->urllistDb->dbp, 0);
	}
	sgFreeAllLists();
	execve(*globalArgv, globalArgv, globalEnvp);
	fprintf(stderr, "error execve: %d\n", errno);
	exit(1);
}

void sgAlarm(int signal)
{
	sig_alrm = 1;
	sgTimeNextEvent();
}

static void resetSquidInfo(struct SquidInfo *s) {
	s->protocol[0]  = '\0';
	s->domain[0]    = '\0';
	s->dot          = 0;
	s->url[0]       = '\0';
	s->orig[0]      = '\0';
	s->surl[0]      = '\0';
	s->furl[0]      = '\0';
	s->strippedurl  = s->surl;
	s->port         = 0;
	s->src[0]       = '\0';
	s->srcDomain[0] = '\0';
	s->ident[0]     = '\0';
	s->method[0]    = '\0';
}

static int parseUrl(char * url, struct SquidInfo *s) {

	char * p = NULL;
	char * d = NULL;
	size_t l = 0;
	size_t n = 0;
	char * domain  = NULL;
	char * sdomain = NULL;

	char pathcp[MAX_BUF];

	memset(pathcp,0,sizeof(pathcp));

	strcpy(s->orig, url);

	/* Now convert url to lowercase chars */
	for (p = url; *p != '\0'; p++)
		*p = tolower(*p);

	if ((p = strstr(url, "://")) == NULL) {
		strcpy(s->protocol, "unknown");
		p = url;
	} else {
		*p = 0;
		strcpy(s->protocol, url);
		p += 3;
	}

	/* p points to the begining of the host part, find first slash that ends it */
	domain = p;
	if ((d = strchr(p,'/')) != NULL) {
		p = d + 1;
		*d = 0;
	} else {
		p = NULL;
	}

	/* p now is either NULL or points to the first char in the path part*/
	if (p != NULL) {
		char * to   = pathcp;
		char * from = p;

		/* skip leading slashes */
		while( *from == '/' )
			from ++;

		while ( *from != 0 && *from != '?' ) {
			/* skip double slashes */
			if (*from == '/' && *(from+1) == '/') {
				from ++;
				continue;
			}
			*to = *from;
			to ++; from ++;
		}

		/* keep the query part when present */
		if ( *from == '?' )
			strcat(to,from);
	}

	/* skip authentication */
	if ((d = strchr(domain, '@')) != NULL) {
		domain = d + 1;
	}

	/* look for a port */
	if ((d = strrchr(domain,':')) != NULL) {
		n = strspn(d + 1, "0123456789");
		if ( *(d + n + 1) == 0 ) {
			s->port = atoi(d + 1);
			*d = 0;
		}
	}

	if (*domain == 0)
		return 0;

	strcpy(s->domain, domain);
	l = strlen(s->domain);

	if (strspn(s->domain, ".0123456789") == l ) {
		/* may be an IPv4 address */
		unsigned char binaddr[sizeof(struct in_addr)];
		int changed = 0;

		if ( inet_pton(AF_INET, s->domain, &binaddr) > 0 ) {
			struct hostent * hp = NULL;

			if (reverselookup) {
				if ((hp = gethostbyaddr( binaddr, sizeof(binaddr), AF_INET )) != NULL) {
					if (strlen(hp->h_name) < MAX_BUF) {
						strcpy(s->domain, hp->h_name);
						changed = 0;
					}
				}
			}

			if (!changed)
				s->dot = 1;
		}
	} else if (s->domain[0] == '[' && s->domain[l- 1] == ']' &&
		   (strspn(s->domain + 1, ":0123456789abcdef") == l - 2)) {
		/* may be an IPv6 address */
		unsigned char binaddr[sizeof(struct in6_addr)];
		int changed = 0;

		s->domain[l-1] = 0;

		if ( inet_pton(AF_INET6, s->domain + 1, &binaddr) > 0 ) {
			struct hostent * hp = NULL;

			if (reverselookup) {
				if ((hp = gethostbyaddr( binaddr, sizeof(binaddr), AF_INET6 )) != NULL) {
					if (strlen(hp->h_name) < MAX_BUF) {
						strcpy(s->domain, hp->h_name);
						changed = 1;
					}
				}
			}

			if (!changed)
				s->dot = 1;
		}

		if (!changed)
			s->domain[l-1] = ']';
	}

	/* strip trailing dot from domain */
	if ((d = s->domain + strlen(s->domain) - 1) >= s->domain) {
		if (*d == '.')
			*d = 0;
	}

	/* strip common host names like www.foo.com or web01.bar.org  */
	sdomain = s->domain;

	if ((domain[0] == 'w' &&  domain[1] == 'w' && domain[2] == 'w') ||
	    (domain[0] == 'w' &&  domain[1] == 'e' && domain[2] == 'b') ||
	    (domain[0] == 'f' &&  domain[1] == 't' && domain[2] == 'p'))
	{
		sdomain += 3;
		while (sdomain[0] >= '0' && sdomain[0] <= '9')
			sdomain ++;

		if (sdomain[0] == '.')
			sdomain ++;
		else
			sdomain = domain;
	}

	if (s->port > 0) {
		sprintf(s->furl, "%s:%d/%s", domain, s->port, pathcp);
		sprintf(s->surl, "%s:%d/%s", sdomain, s->port, pathcp);
	} else {
		sprintf(s->furl, "%s/%s", domain, pathcp);
		sprintf(s->surl, "%s/%s", sdomain, pathcp);
	}

	return 1;
}

/*
 * Parse an external acl helper line.
 *
 * Squid can be configured to pass various formats, we assume something
 * similar to the normal redirector format:
 *
 *   %URI %SRC %LOGIN
 *
 * For example:
 *   external_acl_type foo ttl=60 children=1 %URI %SRC %LOGIN /path/to/sg
 */
int parseAuthzLine(char *line, struct SquidInfo *s)
{
	char * field = NULL;

	sgLogDebug("got authz line %s\n", line);

	resetSquidInfo(s);

	/* get the URL and parse */
	if ((field = strtok(line, "\t ")) == NULL)
		return 0;

	HTUnEscape(field);
	if (!parseUrl(field,s)) {
		return 0;
	}

	/* get the source address and parse */
	if ((field = strtok(NULL, " \t\n")) == NULL)
		return 0;

	HTUnEscape(field);	/* just in case, IPs should not need escaping */
	strcpy(s->src, field);

	/* get the login and parse */
	if ((field = strtok(NULL, " \t\n")) == NULL)
		return 0;

	HTUnEscape(field);

	strcpy(s->ident, field);

	for (field = s->ident; *field != '\0'; field++) /* convert ident to lowercase chars */
		*field = tolower(*field);

	sgLogDebug( "parsed authz line: furl='%s' domain='%s' surl='%s' src=%s ident='%s'\n",
			s->furl, s->domain, s->surl, s->src, s->ident );

	return 1;
}

/*
 * Parse a redirector input line, format is:
 *
 *   URL ip-address/fqdn ident method
 *
 * for example
 *    http://www.example.com/page1.html 192.168.2.3/- andi GET
 */

int parseLine(char *line, struct SquidInfo *s)
{
	char * field = NULL;
	char * p = NULL;

	sgLogDebug("got redirector line %s\n", line);

	resetSquidInfo(s);

	/* get the URL and parse */
	if ((field = strtok(line, "\t ")) == NULL)
		return 0;

	HTUnEscape(field);
	if (!parseUrl(field,s)) {
		return 0;
	}

	/* get the source address and parse */
	if ((field = strtok(NULL, " \t\n")) == NULL)
		return 0;

	if ((p = strchr(field, '/')) != NULL) {
		*p = 0;
		strcpy(s->src, field);
		strcpy(s->srcDomain,p + 1);
		if (s->srcDomain[0] == '-' && s->srcDomain[1] == 0)
			s->srcDomain[0] = 0;

	} else {
		strcpy(s->src, field);
	}

	/* get the identity */
	if ((field = strtok(NULL, " \t\n")) == NULL)
		return 0;

	if (strcmp(field, "-")) {
		strcpy(s->ident, field);
		for (p = s->ident; *p != '\0'; p++) /* convert ident to lowercase chars */
			*p = tolower(*p);
	} else {
		s->ident[0] = '\0';
	}

	/* get the method */
	if ((field = strtok(NULL, " \t\n")) == NULL)
		return 0;

	strcpy(s->method, field);
	if (s->method[0] == '\0')
		return 0;

	sgLogDebug( "parsed redirector line: furl='%s' domain='%s' surl='%s' src=%s ident='%s'\n",
			s->furl, s->domain, s->surl, s->src, s->ident );

	return 1;
}

char *sgStripUrl(char *url)
{
	static char newurl[MAX_BUF];
	char *p, *d = NULL, *a = NULL, *e = NULL;
	p = url;
	d = strchr(p, '/');     /* find domain end */
	e = d;
	a = strchr(p, '@');     /* find auth  */
	if (a != NULL && (a < d || d == NULL))
		p = a + 1;
	a = strchr(p, ':'); /* find port */;
	if (a != NULL && (a < d || d == NULL))
		e = a;
	if (e == NULL) {
		strcpy(newurl, p);
	} else {
		strncpy(newurl, p, e - p);
		*(newurl + (e - p)) = '\0';
	}
	if (d != NULL)
		strcat(newurl, d);
	return newurl;
}

/*
 * returns a pointer to the domain part of a fully-qualified  hostname
 * so www.abc.xyz.dom/index.html -> xyz.dom/index.html
 */

char *sgSkipHostPart(char *domain)
{
	char *p = domain, *d1 = NULL, *d2 = NULL, *path = NULL;
	if ((path = (char *)strchr(p, '/')) == NULL)
		path = domain;
	while ((p = (char *)strchr(p, '.')) != NULL) {
		if (p > path && path != domain)
			break;
		d2 = d1;
		d1 = p;
		p++;
	}
	if (d2 != NULL)
		return d2 + 1;
	return domain;
}

/*
 *
 * checks the vality of an dotted address.
 *
 */

ulong *sgConvDot(char *dot)
{
	static unsigned long ipaddr = 0;
	int octet;
	char *s = dot, *t;
	int shift = 24;
	ipaddr = 0;
	while (*s) {
		t = s;
		if (!isdigit(*t))
			return NULL;
		while (isdigit(*t))
			++t;
		if (*t == '.')
			*t++ = 0;
		else if (*t)
			return NULL;
		if (shift < 0)
			return NULL;
		octet = atoi(s);
		if (octet < 0 || octet > 255)
			return NULL;
		ipaddr |= octet << shift;
		s = t;
		shift -= 8;
	}
	return &ipaddr;
}


/*
 * Reverses cmp of strings
 */

int sgStrRcmp(char *a, char *b)
{
	char *a1 = (char *)strchr(a, '\0');
	char *b1 = (char *)strchr(b, '\0');
	while (*a1 == *b1) {
		if (b1 == b || a1 == a)
			break;
		a1--; b1--;
	}
	if (a1 == a && b1 == b)
		return *a1 - *b1;
	if (a1 == a)
		return -1;
	if (b1 == b)
		return 1;
	return *a1 - *b1;
}

int sgStrRncmp(char *a, char *b, int blen)
{
	char *a1 = (char *)strchr(a, '\0');
	char *b1 = (char *)strchr(b, '\0');
	while (*a1 == *b1 && blen > 0) {
		if (b1 == b || a1 == a)
			break;
		a1--; b1--; blen--;
	}
	if (a1 == a && b1 == b)
		return *a1 - *b1;
	if (blen == 0)
		return *a1 - *b1;
	if (a1 == a)
		return -1;
	if (b1 == b)
		return 1;
	return *a1 - *b1;
}

/*
 *
 * sgDomStrRncmp checks if B is equal to or a subdomain of A
 *
 */


int sgDomStrRcmp(char *p1, char *p2)
{
	char *p11 = (char *)strchr(p1, '\0');
	char *p22 = (char *)strchr(p2, '\0');
	for (; p11 >= p1 && p22 >= p2 && *p11 == *p22; p11--, p22--) ;
	if (p11 < p1 && p22 < p2)
		return 0;
	if (p22 < p2)
		return -*p11;
	if (p11 < p1 && *p22 == '.')
		return 0;
	return *p11 - *p22;
}

/*
 *
 * Regexp functions
 *
 */

struct sgRegExp *sgNewPatternBuffer(char *pattern, int flags)
{
	regex_t *compiled = sgMalloc(sizeof(regex_t));
	struct sgRegExp *regexp;
	regexp = sgMalloc(sizeof(struct sgRegExp));
	regexp->pattern = sgMalloc(strlen(pattern) + 1);
	strcpy(regexp->pattern, pattern);
	regexp->error = 0;
	regexp->next = NULL;
	regexp->flags = flags;
	regexp->error = regcomp(compiled, pattern, flags);
	regexp->compiled = compiled;
	return regexp;
}

/*
 * Deletes the buffer memory, so save the next pointer first before
 * calling this function.
 */
void sgFreePatternBuffer(struct sgRegExp *regexp)
{
	sgFree(regexp->pattern);
	sgFree(regexp->compiled);
	sgFree(regexp);
}

char *sgRegExpSubst(struct sgRegExp *regexp, char *pattern)
{
	struct sgRegExp *re;
	regmatch_t pm[10];
	static char newstring[MAX_BUF];
	char *result = NULL, *p;
	int substlen;
	*newstring = '\0';
	for (re = regexp; re != NULL; re = re->next) {
		if (regexec(re->compiled, pattern, sizeof(pm) / sizeof(pm[0]), pm, 0) != 0) {
			result = NULL;
		} else {
			substlen = strlen(re->substitute);
			if (re->httpcode != NULL)
				strcpy(newstring, re->httpcode);
			else
				*newstring = '\0';
			p = newstring;
			do {
				if ((p - newstring) + pm[0].rm_so >= MAX_BUF)
					break;
				p = strncat(newstring, pattern, pm[0].rm_so);
				{
					char *p_cur;
					char *p_next;

					for (p_next = p_cur = re->substitute;
					     p_next < (re->substitute + substlen);
					     p_next++) {
						if (*p_next == '\\') {
							if (p_cur < p_next) {
								if (((p - newstring) + (p_next - p_cur)) >= MAX_BUF)
									goto err;
								p = strncat(newstring, p_cur, p_next - p_cur);
							}
							p_next++;
							if (p_next < (re->substitute + substlen)
							    && '0' <= *p_next && *p_next <= '9') {
								int i = *p_next - '0';
								if ((p - newstring) + (pm[i].rm_eo - pm[i].rm_so) >= MAX_BUF)
									goto err;
								p = strncat(newstring, pattern + pm[i].rm_so, pm[i].rm_eo - pm[i].rm_so);
							} else {
								if ((p - newstring + 1) >= MAX_BUF)
									goto err;
								p = strncat(newstring, p_next, 1);
							}
							p_cur = p_next + 1;
						} else if (*p_next == '&') {
							if (p_cur < p_next) {
								if (((p - newstring) + (p_next - p_cur)) >= MAX_BUF)
									goto err;
								p = strncat(newstring, p_cur, p_next - p_cur);
							}
							if (((p - newstring) + (pm[0].rm_eo - pm[0].rm_so)) >= MAX_BUF)
								goto err;
							p = strncat(newstring, pattern + pm[0].rm_so, pm[0].rm_eo - pm[0].rm_so);
							p_cur = p_next + 1;
						}
					}
					if (p_cur < p_next) {
						if (((p - newstring) + (p_next - p_cur)) >= MAX_BUF)
							goto err;
						p = strncat(newstring, p_cur, p_next - p_cur);
					}
				}
				pattern = pattern + pm[0].rm_eo;
			} while (regexec(re->compiled, pattern, sizeof(pm) / sizeof(pm[0]), pm, REG_NOTBOL) == 0 && re->global);
			if ((p - newstring) + strlen(pattern) <= MAX_BUF)
				p = strcat(newstring, pattern);
			result = newstring;
			break;
		}
	}
err:
	return result;
}

/*
 *
 *
 *
 */

char *sgParseRedirect(char *redirect, struct SquidInfo *req, struct Acl *acl, struct AclDest *aclpass)
{
	static char buf[MAX_BUF + MAX_BUF];
	char *p = redirect, *q = NULL, *t = NULL;
	struct Source *s = lastActiveSource;
	*buf = '\0';
	if (aclpass == NULL)
		aclpass = defaultAcl->pass;
	while ((p = strchr(p, '%')) != NULL) {
		if (q == NULL) {
			strncpy(buf, redirect, p - redirect);
			buf[p - redirect] = '\0';
		} else {
			strncat(buf, q, p - q);
		}
		if (p == NULL)
			break;
		switch (*(p + 1)) {
		case 'a': /* Source Address */
			strcat(buf, req->src);
			p++;
			break;
		case 'i': /* Source User Ident */
			if (!strcmp(req->ident, "-"))
				strcat(buf, "unknown");
			else
				strcat(buf, req->ident);
			p++;
			break;
		case 'q': /* userquota info */
			if (s != NULL && s->userquota.seconds != 0 && strcmp(req->ident, "-")) {
				struct UserInfo *userquota;
				if (defined(s->userDb, req->ident, (char **)&userquota) == 1) {
					char qbuf[150];
					sprintf(qbuf, "%d-%ld-%d-%ld-%ld-%d",
					        s->userquota.renew,
					        s->userquota.seconds,
					        userquota->status,
					        userquota->time,
					        userquota->last,
					        userquota->consumed);
					strcat(buf, qbuf);
				} else {
					strcat(buf, "noquota");
				}
			} else {
				strcat(buf, "noquota");
			}
		case 'n': /* Source Domain Name */
			if (!strcmp(req->srcDomain, "-"))
				strcat(buf, "unknown");
			else
				strcat(buf, req->srcDomain);
			p++;
			break;
		case 'p': /* The url path */
			if ((t = strstr(req->orig, "//")) != NULL) {
				t += 2;
				if ((t = strchr(t, '/')) != NULL)
					strcat(buf, ++t);
			}
			p++;
			break;
		case 'f': /* The url file */
			if ((t = strrchr(req->orig, '/')) != NULL) {
				t++;
				strcat(buf, t);
			}
			p++;
			break;
		case 's': /* Source Class Matched */
			if (acl->source == NULL || acl->source->name == NULL)
				strcat(buf, "default");
			else
				strcat(buf, acl->source->name);
			p++;
			break;
		case 't': /* Target Class Matched */
			if (aclpass == NULL) {
				strcat(buf, "unknown");
			} else if (aclpass->name == NULL) {
				if (aclpass->type == ACL_TYPE_INADDR)
					strcat(buf, "in-addr");
				else if (aclpass->type == ACL_TYPE_TERMINATOR)
					strcat(buf, "none");
				else
					strcat(buf, "unknown");
			} else {
				strcat(buf, aclpass->name);
			}
			p++;
			break;
		case 'u': /* Requested URL */
			strncat(buf, req->orig, 2048);
			p++;
			break;
		default:
			strcat(buf, "%")
			;
		}
		p++;
		q = p;
	}
	if (buf[0] == '\0')
		q = redirect;
	else
		q = buf;
	return q;
}


/*
 * converts yyyy.mm.ddTHH:MM:SS to seconds since EPOC
 */

time_t iso2sec(char *date)
{
	struct tm *t;
	int y, m, d, H, M, S;
	t = sgMalloc(sizeof(struct tm));
	sscanf(date, "%4d%*[.-]%2d%*[.-]%2d%*[T]%2d%*[:-]%2d%*[:-]%2d",
	       &y, &m, &d, &H, &M, &S);
	m--;
	y = y - 1900;
	if (y < 0 || m < 0 || m > 11 || d < 1 || d > 31 || H < 0 || H > 23
	    || M < 0 || M > 59 || S < 0 || S > 59)
		return (time_t)-1;
	t->tm_year = y;
	t->tm_mon = m;
	t->tm_mday = d;
	t->tm_hour = H;
	t->tm_min = M;
	t->tm_sec = S;
	return (time_t)mktime(t);
}

/*
 * converts yyyy.mm.dd to seconds since EPOC
 */

time_t date2sec(char *date)
{
	struct tm *t;
	int y, m, d;
	t = sgMalloc(sizeof(struct tm));
	sscanf(date, "%4d%*[.-]%2d%*[.-]%2d", &y, &m, &d);
	m--;
	y = y - 1900;
	if (y < 0 || m < 0 || m > 11 || d < 1 || d > 31)
		return (time_t)-1;
	t->tm_year = y;
	t->tm_mon = m;
	t->tm_mday = d;
	return (time_t)mktime(t);
}

char *niso(time_t t)
{
	static char buf[20];
	time_t tp;
	struct tm *lc;
	if (t == 0)
		tp = time(NULL) + globalDebugTimeDelta;
	else
		tp = t;
	lc = localtime(&tp);
	sprintf(buf, "%04d-%02d-%02d %02d:%02d:%02d", lc->tm_year + 1900, lc->tm_mon + 1,
		lc->tm_mday, lc->tm_hour, lc->tm_min, lc->tm_sec);
	return buf;
}

struct UserInfo *setuserinfo()
{
	static struct UserInfo uq;

	uq.status = 0;
	uq.time = 0;
	uq.consumed = 0;
	uq.last = 0;
#ifdef HAVE_LIBLDAP
	uq.ldapuser = 0;
	uq.found = 0;
	uq.cachetime = 0;
#endif
	return &uq;
}

#ifdef HAVE_LIBLDAP
struct IpInfo *setipinfo()
{
	static struct IpInfo uq;

	uq.status = 0;
	uq.time = 0;
	uq.consumed = 0;
	uq.last = 0;
	uq.ldapip = 0;
	uq.found = 0;
	uq.cachetime = 0;
	return &uq;
}
#endif
